### PR TITLE
Musca dome robustness

### PR DESCRIPTION
**Pull request by danjampro**
A few changes (tested on Huntsman) to make the dome more robust.

### CODE REVIEW

File: src/huntsman/pocs/dome/musca.py [hunk 1]
```diff
@@ -3,11 +3,9 @@
 import time
 from threading import Lock
 
-import astropy.units as u
+from astropy import units as u
 
-from panoptes.utils import CountdownTimer
-from panoptes.utils import error
-from panoptes.utils import get_quantity_value
+from panoptes.utils import error, get_quantity_value
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py:7]: We *really* need a set style policy...

File: src/huntsman/pocs/dome/musca.py [hunk 2]
```diff
@@ -62,23 +59,23 @@ class HuntsmanDome(AbstractSerialDome):
     # V, so we don't open if less than this or CLose immediately if we go less than this
     MIN_OPERATING_VOLTAGE = 12.
 
-    def __init__(self, command_delay=10, max_status_attempts=10, *args, **kwargs):
+    def __init__(self, command_delay=1, max_status_attempts=10, move_timeout=60, *args, **kwargs):
         """
         Args:
             command_delay (float, optional): Wait this long in seconds before allowing next command
                 due to slow musica CPU. Default 10s.
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]:
```suggestion
                due to slow musca CPU. Default 1s.
```

File: src/huntsman/pocs/dome/musca.py [hunk 3]
```diff
@@ -110,28 +107,14 @@ def status(self):
...
         TODO: Auto-update this every ~60s
         """
-        if self._status_timer is None:
-            update_status = True
-        else:
-            update_status = self._status_timer.expired()
-
-        if update_status:
+        with self._command_lock:  # Make status call thread-safe
             for i in range(1, self._max_status_attempts + 1):
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]:
```suggestion
            for i in range(self._max_status_attempts):
```

File: src/huntsman/pocs/dome/musca.py [hunk 4]
```diff
@@ -62,23 +59,23 @@ class HuntsmanDome(AbstractSerialDome):
...
 
-    def __init__(self, command_delay=10, max_status_attempts=10, *args, **kwargs):
+    def __init__(self, command_delay=1, max_status_attempts=10, move_timeout=60, *args, **kwargs):
         """
         Args:
             command_delay (float, optional): Wait this long in seconds before allowing next command
                 due to slow musica CPU. Default 10s.
             max_status_attempts (int, optional): If status fails, retry this many times before
                 raising a PanError. Default: 10.
+            move_timeout (u.Quantity, optional): The dome shutter movement timeout. Default 60s.
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]: Seems like this should be `shutter_timeout`?

File: src/huntsman/pocs/dome/musca.py [hunk 5]
```diff
@@ -234,13 +200,12 @@ def __del__(self):
 
     def _write_musca(self, cmd, log_message=None):
         """Wait for the command lock then write command to serial bluetooth device musca."""
-        with self._command_lock:
```
Comment (fergusL) [inline src/huntsman/pocs/dome/musca.py:237]: why remove the lock here?

File: src/huntsman/pocs/dome/musca.py [hunk 6]
```diff
@@ -150,8 +133,8 @@ def open(self):
                 'Dome shutter battery voltage too low to open: {!r}', v)
             return False
 
-        self._write_musca(Protocol.OPEN_DOME, 'Opening dome shutter')
-        time.sleep(HuntsmanDome.MOVE_TIMEOUT)
+        self._write_musca(Protocol.OPEN_DOME, 'Opening dome shutter.')
+        time.sleep(self._move_timeout)
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]: Is this just to wait until it's open?  Seems better to loop and check the `v == Protocol.OPEN` below and set a `max_wait` timer.

**Review by fergusL — COMMENTED** [review]
Was there a reason to remove the lock on `write_musca`?

File: src/huntsman/pocs/dome/musca.py (hunk 5 above)
Comment (wtgee) [reply src/huntsman/pocs/dome/musca.py:237]: Yeah, the built in one second delay after write versus a lock seems odd.

**Review by wtgee — APPROVED** [review]
This isn't really my favorite, but I'll approve.

Ideally we should really try to find the musca source to see what is going on instead of building weird non-intuitive hacks into this code. But I don't think we can update musca until someone is on-site again 😦

File: src/huntsman/pocs/dome/musca.py (hunk 5 above)
Comment (danjampro) [reply src/huntsman/pocs/dome/musca.py:237]: I've moved the lock so we are blocking for the whole write/read for the `status` property. It is also implemented on write for `open` and `close` methods.

File: src/huntsman/pocs/dome/musca.py [hunk 7]
```diff
@@ -141,106 +140,88 @@ def open(self):
         Boolean
             True if Opened, False if it did not Open.
         """
+        self._keep_open = True
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]: Should this be set after the two exception checks below?

File: src/huntsman/pocs/dome/musca.py (hunk 3 above)
Comment (wtgee) [reply src/huntsman/pocs/dome/musca.py]: You either need to change this as suggested (which makes more sense), or change the `{i+1}` below.

File: src/huntsman/pocs/dome/musca.py [hunk 8]
```diff
@@ -104,34 +109,28 @@ def door_closed(self):
...
+        """A dictionary containing all status info for dome. """
+        with self._command_lock:  # Make status call thread-safe
             for i in range(1, self._max_status_attempts + 1):
+                try:
+                    status = self._get_status_dict()
+                    status["dome_thread_running"] = self._dome_thread.is_alive()
+                    status["keep_shutter_open"] = self._keep_open
+                    return status
+                except Exception as err:
+                    self.logger.warning(f"Retrying dome status: {i+1} of"
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]: See note above.

File: src/huntsman/pocs/dome/musca.py [hunk 9]
```diff
@@ -104,34 +109,28 @@ def door_closed(self):
...
 
-        if update_status:
+    @property
+    def status(self):
+        """A dictionary containing all status info for dome. """
+        with self._command_lock:  # Make status call thread-safe
             for i in range(1, self._max_status_attempts + 1):
+                try:
+                    status = self._get_status_dict()
+                    status["dome_thread_running"] = self._dome_thread.is_alive()
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]: This is the dome status, so `dome_thread_running` isn't informative to the user (or is misinformative)

```suggestion
                    status["status_thread_running"] = self._dome_thread.is_alive()
```

File: src/huntsman/pocs/dome/musca.py [hunk 10]
```diff
@@ -104,34 +109,28 @@ def door_closed(self):
...
+                    status["keep_shutter_open"] = self._keep_open
+                    return status
+                except Exception as err:
+                    self.logger.warning(f"Retrying dome status: {i+1} of"
+                                        f" {self._max_status_attempts}: {err}")
```
Comment (wtgee) [inline src/huntsman/pocs/dome/musca.py]: Exception look better with the `!r` modifier.

```suggestion
                                        f" {self._max_status_attempts}: {err!r}")
```

**Review by wtgee — APPROVED** [review]

**Pull request merged**